src/mlpack/methods/boostmetric/boostmetric_impl.hpp

```diff
@@ -0,0 +1,179 @@
+
+  return sum;
+}
+
+template<typename MetricType>
+void BOOSTMETRIC<MetricType>::LearnDistance(arma::mat& outputMatrix)
+{
+  Timer::Start("boostmetric");
+
+  // See if we were passed an initialized matrix.
+  if ((outputMatrix.n_rows != dataset.n_rows) ||
+      (outputMatrix.n_cols != dataset.n_rows))
+  {
+    outputMatrix.eye(dataset.n_rows, dataset.n_rows);
+  }
+
+  // Compute triplets.
+  arma::Mat<size_t> triplets;
+  constraint.Triplets(triplets, dataset, labels);
+
+  size_t N = triplets.n_cols;
+  size_t dim = dataset.n_rows;
+
+  // Initialize u.
+  arma::vec u;
+  u.ones(N);
+
+  // Normalize u such that sum(u_i) = 1
+  u = u / N;
+
+  arma::cube A(dim, dim, N);
+
+  // Initialize A (coefficient matrix of difference between distance
+  // of impostors & target neighbors).
+  for (size_t i = 0; i < N; i++)
+  {
+    A.slice(i) = (dataset.col(triplets(0, i)) - dataset.col(triplets(2, i))) *
+        arma::trans(dataset.col(triplets(0, i)) - dataset.col(triplets(2, i)))
+        -
+        (dataset.col(triplets(0, i)) - dataset.col(triplets(1, i))) *
+        arma::trans(dataset.col(triplets(0, i)) - dataset.col(triplets(1, i)));
+  }
+
+  // Declare weighted sum of Ar matrices.
+  arma::mat Acap;
+
+  // Now iterate!
+  for (size_t i = 0; i < maxIter; i++)
+  {
+    // Initialize Acap.
+    Acap.zeros(dim, dim);
+    for (size_t j = 0; j < N; j++)
+      Acap += u(j) * A.slice(j);
+
+    arma::vec eigval;
+    arma::mat eigvec;
+    arma::eig_sym(eigval, eigvec, (Acap + trans(Acap) / 2));
+
+    // Get maximum eigvalue.
+    double maxEig = arma::max(eigval);
+
+    if (maxEig < tolerance)
+    {
+      Log::Info << "BoostMetric: minimized within tolerance " << tolerance
+            << "; " << "terminating optimization." << std::endl;
+      break;
+    }
+
+    // Output current maximum eigen value.
+    Log::Info << "BoostMetric: iteration " << i << ", maximum eigen value "
+        <<  maxEig << "." << std::endl;
+
+    // Get index of maximum element.
+    arma::uvec maxIndex = arma::find(eigval == maxEig);
+
+    arma::mat Z = eigvec.col(maxIndex(0)) * arma::trans(eigvec.col(maxIndex(0)));
+
+    // Initialize H vector by computing inner product. Hr = <Ar, Z>
+    arma::vec H(N);
+    for (size_t j = 0; j < N; j++)
+      H(j) = innerProduct(A.slice(j), Z);
+
+    // Binary Search for weight calculation.
+    double w;
+    while (true)
+    {
+      // Calculate mid value.
+      w = (wHigh + wLow) * 0.5;
+
+      double lhs = 0;
+      for (size_t j = 0; j < N; j++)
+        lhs += (H(j) - tolerance) * std::exp(-w * H(j)) * u(j);
+
+      // Update search direction.
+      if (lhs > 0)
+        wLow = w;
+      else
+        wHigh = w;
+
+      // Terminate search.
+      if (wHigh - wLow < wTolerance || std::abs(lhs) < wTolerance)
+        break;
+    }
+
+    // Update u.
+    for (size_t j = 0; j < N; j++)
+      u(j) = u(j) * std::exp(-H(j) * w);
+
+    // Normalize u so that sum(u_i) = 1.
+    u /= arma::sum(u);
```

Comment:
Do we end up with cases where `u` is either extremely small or 0?  In that case we could avoid adding it to the sum `Acap`.

Reply:
Here's the simulation supporting this - https://gist.github.com/manish7294/01bb5b6d2f5c4cbdb60dc5af541e39dc
Let me know your thoughts on this.

Reply:
I see, it looks like this could be a good pruning opportunity if we have time (but it's not clear if this will provide much speedup to me, I'm not sure how much computation we save).